extension.mk

```diff
@@ -0,0 +1,256 @@
+TEST_C_FILES=$(TEST_TRANS_XC_FILES:.xc=.c)
+# Object files that should be generated
+EXAMPLE_OBJECTS=$(EXAMPLE_XC_FILES:.xc=.o)
+TEST_OBJECTS=$(TEST_TRANS_XC_FILES:.xc=.o)
+# Executables that should be generated
+EXAMPLE_EXECUTABLES=$(EXAMPLE_XC_FILES:.xc=.out)
+TEST_EXECUTABLES=$(TEST_TRANS_XC_FILES:.xc=.out)
+# Evaluation marker files that should be generated
+EXAMPLE_TESTS=$(EXAMPLE_XC_FILES:.xc=.test)
+TEST_TESTS=$(TEST_XC_FILES:.xc=.test)
+
+# Flags passed to the C compiler, e.g. to enable various compiler extensions
+override CFLAGS+=-g
+# Flags passed to the linker
+override LDFLAGS+=-g
+
+all: build libraries examples analyses test
+
+build: $(ARTIFACT_JAR) compiler.jar
+
+ifeq ($(LIB_NAME),)
+  libraries:
+else
+  libraries: $(LIB_C_GEN_FILES) $(LIB_OBJECTS) $(SHARED_LIBRARY) $(STATIC_LIBRARY)
+endif
+
+examples: $(EXAMPLE_C_FILES) $(EXAMPLE_OBJECTS) $(EXAMPLE_EXECUTABLES) $(EXAMPLE_TESTS)
+test: $(TEST_C_FILES) $(TEST_OBJECTS) $(TEST_EXECUTABLES) $(TEST_TESTS)
+check: test
+
+analyses: mda mwda
+mda: mda.test
+mwda: mwda.test
+
+# This file is generated below, containing dependency information for included extensions,
+# and also setting variables to pass appropriate flags.
+include depends.mk
+
+generated bin lib:
+	mkdir -p $@
+
+$(ABLEC_JAR): $(shell find $(ABLEC_BASE)/grammars/ -name *.sv -print0 | xargs -0)
+	cd $(ABLEC_BASE) && ./build $(SVFLAGS)
+
+$(EXTS_BASE)/%.jar:
+	$(MAKE) -C $(dir $@) $(notdir $@)
+
+ifdef USE_CUSTOM_SILVER
+# Note that $(DEP_JARS) are order-only dependencies, to avoid expensive rebuilds.
+# If dependency extension syntax changes, this may require `make depclean` to be reflected.
+silver-compiler.jar: | $(DEP_JARS) generated
+	silver -o $@ $(SVFLAGS) $(EXT_GRAMMAR):artifacts:silver_compiler
+endif
+
+$(ARTIFACT_JAR): $(GRAMMAR_SOURCES) $(DEP_JARS) $(SV_COMPILER_JAR) | generated
+	$(SILVER) -o $@ $(SVFLAGS) $(EXT_GRAMMAR)
+
+compiler.jar: ${ARTIFACT_JAR} $(GRAMMAR_SOURCES) $(DEP_JARS) $(SV_COMPILER_JAR) | generated
+# TODO: Shouldn't need to use the extended Silver here?
+	$(SILVER) -o $@ -I $(ARTIFACT_JAR) $(SVFLAGS) $(EXT_GRAMMAR):artifacts:compiler
+
+mda.test: $(ARTIFACT_JAR) $(DEP_JARS) $(SV_COMPILER_JAR) | generated
+# TODO: Shouldn't need to use the extended Silver here?
+	$(SILVER) --dont-translate --build-xml-location build_mda.xml -I $(ARTIFACT_JAR) $(SVFLAGS) $(EXT_GRAMMAR):artifacts:mda_test
+	touch $@
+
+mwda.test: $(GRAMMAR_SOURCES) $(DEP_JARS) $(SV_COMPILER_JAR) | generated
+	$(SILVER) --dont-translate --mwda --clean --build-xml-location build_mwda.xml $(SVFLAGS) $(EXT_GRAMMAR)
+	touch $@
+
+bin/%.c: src/%.xc $(XC_INCLUDE_SOURCES) compiler.jar | bin
+	java $(JAVAFLAGS) -jar compiler.jar $< $(LIB_XCFLAGS)
+	mv src/$*.c src/$*.i bin
+
+%.c: %.xc $(XC_INCLUDE_SOURCES) compiler.jar
+	java $(JAVAFLAGS) -jar compiler.jar $< $(XCFLAGS)
+
+bin/%.o: src/%.c $(XC_INCLUDE_SOURCES) | bin
+	$(CC) -c $(LIB_CPPFLAGS) $(LIB_CFLAGS) $< -o $@
+
+bin/%.o: bin/%.c | bin
+	$(CC) -c $(LIB_CFLAGS) $< -o $@
```

Comment:
Those are the generated C files for building an extension library, coming from .xc files in src/.  I don't want to put them under src/ since there can also be source c files there.  Do you have a problem with putting them in bin/?

Reply:
I'm not thrilled with it; how do you feel about `build` or `tmp` for these intermediate artifacts? (I tend to think of `bin` as being intended so you can do `install -D "$DESTDIR$PREFIX/bin" bin/*`)

Reply:
I'm putting object files there, too.  I suppose `build` is fine.

Reply:
Renaming `generated/` to `build/` as we discussed offline actually seems like too much of a pain, e.g. the Jenkins lib assumes Silver gen file folders are always named `generated/`.  So I'll just make it put the library files in `generated/` instead of `bin/`.